Add an explainer for subresource signed exchanges (#542)
WICG/webpackage: explainers/signed-exchange-subresource-subtitution-explainer.md

> # Signed Exchange subresource substitution

Author: Merged.

## Introduction

> We want to introduce a new **rel=allowed-alt-sxg** link HTTP header which will

Author: Done

the user must wait for the JS file to be fetched from the server to read the
article.

If the feed site knows that the article is depending on the JS file, the site
can insert a link element (`<link rel="prefetch"
href="https://cdn.publisher.example/lib.js" as="script">`) to prefetch the JS
file while the user is browsing the feed site. But this is problematic from the
privacy point of view. And also this doesn’t work if the UA is using
[origin isolated HTTPCache mechanism](http://sirdarckcat.blogspot.com/2019/03/http-cache-cross-site-leaks.html).

Signed Exchange has solved this
[privacy-preserving prefetching](https://wicg.github.io/webpackage/draft-yasskin-webpackage-use-cases.html#private-prefetch)
problem for main resources. If the publisher is providing the article in signed
exchange format (article.html.sxg), the UA can prefetch the signed exchange from
the feed site’s own server while the user is browsing the feed site. But there
is no way to prefetch subresources in a privacy-preserving manner yet.

Our proposal can solve this problem:
1. The publisher provides the script file in signed exchange format (lib.js.sxg)

> along with its [header integrity value](#header-integrity-of-signed-exchange).

Reviewer: The publisher doesn't have to independently provide the header integrity value, right? That's computable from the SXG itself?

1. The UA prefetches the signed exchange of the script from the feed site’s own
server while the user is browsing the feed site.

> 1. The HTTP response of the article from the publisher's server (https://publisher.example/article.html) has an allowed-alt-sxg link header to declare

Reviewer: ```suggestion
1. The signed HTTP response inside the signed exchange (https://publisher.example/article.html) has an `allowed-alt-sxg` link header to declare
```

> that the UA can load the script from the prefetched signed exchange.

Reviewer: ```suggestion
exactly what version of the subresource is compatible with this main resource.
```

   ```
   Link: <https://cdn.publisher.example/lib.js>;
         rel="allowed-alt-sxg";
         header-integrity="sha256-..."
   ```

>    So, the UA can load the JS file from the prefetched signed exchange. If the

Reviewer: I think this paragraph can go away?

   publisher is providing the article in signed exchange format
   (article.html.sxg), the allowed-alt-sxg link header is in the inner response
   headers of the signed exchange.

> ### Reading article while offline

Reviewer: This is probably better solved by bundles.

If the publisher is providing the main resource and necessary subresources in
signed exchange format, the user of the news feed site can read the article even
while offline. This can be done using Service Worker.
1. The Service Worker of the news feed site fetches the main resource signed
exchange and subresources signed exchanges (maybe using
[Background Fetch](https://wicg.github.io/background-fetch/)) while online, and
stores them to the CacheStorage of the feed site origin.
1. While the user is browsing the feed site, the site inserts prefetch link
elements to prefetch the signed exchanges. The service worker responds to the
fetch events of the prefetch with the signed exchanges in the CacheStorage.
1. When the user clicks a link to the article’s signed exchange, the content and
all necessary subresources are loaded from the cached signed exchange. So the
user can read the article while offline.

# Proposal

> 1. Introduce SignedExchangeCache and PrefetchedSignedExchangeCache which are

Reviewer: These seem like Chromium-internal concepts rather than things the TAG needs to think about.

prefetched from the Document. PrefetchedSignedExchangeCache keeps the signed
exchanges which were passed from the referrer Document which triggered the
navigation to the current Document.
1. While processing
"[prefetch](https://html.spec.whatwg.org/multipage/links.html#link-type-prefetch)"
link (eg: `<link rel="prefetch"
href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`):
   - If succeeded prefetching the signed exchange (including merkle integrity
   check of the body), stores the parsedExchange which is the result of
   [parsing a signed exchange](https://wicg.github.io/webpackage/loading.html#ref-for-parsing-a-signed-exchange)
   (= inner request URL and inner response) to the SignedExchangeCache.
1. While
[navigating across documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents),
copy the parsedExchanges in the SignedExchangeCache of the source document to
the PrefetchedSignedExchangeCache of the target document. This is intended to
provide a way to pass the cached signed exchange across origins even if the UA
is using origin isolated HTTPCache mechanism.

> 1. While processing

Reviewer: This seems very similar to other explainer's algorithm. Maybe add anything that's missing there and remove this one?

Author: Merged

[preload](https://html.spec.whatwg.org/multipage/links.html#link-type-preload)
link HTTP headers (eg: Link: <https://cdn.publisher.example/lib.js>;
rel="preload"; as="script"):
   - Check whether matching "allowed-alt-sxg" link HTTP header
   (`Link: <https://cdn.publisher.example/lib.js>;rel="allowed-alt-sxg";header-integrity="sha256-.."`)
   exists or not. (Note that if the allowed-alt-sxg link HTTP header has
   variants and variant-key attributes, the UA must execute the algorithm
   written in
   [HTTP Representation Variants spec](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html)
   to find the matching header.)
   - Check whether all preload links which have matching allowed-alt-sxg link
   header have matching (url and header-integrity) parsedExchange in
   PrefetchedSignedExchangeCache. If the check passes, set request's stashed
   exchange to the parsedExchange so the resource will be loaded from the cached
   signed exchange.
   - After processing the link HTTP headers, clears the
   PrefetchedSignedExchangeCache.

# Detailed design discussion
## Header integrity of signed exchange

> We use the SHA256 hash value of *signedHeaders* in

Reviewer: ```suggestion
The `header-integrity` parameter of the `allowed-alt-sxg` link is the SHA256 hash value of the *signedHeaders* value from the
```

[application/signed-exchange format](https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange)
for integrity checking. This signedHeaders is *"the canonical serialization of
the CBOR representation of the response headers of the exchange represented by
the application/signed-exchange resource, excluding the Signature header
field"*. So this value doesn’t change even if the publisher signs the content

> again or changes the signing key. This header-integrity value can verify the

Reviewer: ```suggestion
again or changes the signing key. This header-integrity value also guarantees the content body hasn't changed, because the signed headers are required to include a `Digest` header with a hash of the content body.
```

integrity of both the response header and the content body, because signed
exchange’s response header must have digest header.

The UA needs to check that this value of the prefetched subresource signed
exchange is same as the header-integrity attribute of allowed-alt-sxg link
header in the response from the publisher. This is intended to prevent the
distributor from encoding a tracking ID into the subresource signed exchange.

We can’t use the
[SRI’s integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)

> for this purpose, because SRI’s integrity can be used only for verifying the

Reviewer: ```suggestion
for this purpose, because SRI’s integrity only covers the content body and not any of the headers.
```

integrity of the content body. So if the UA use the SRI’s integrity value in
allowed-alt-sxg link header, we can use the subresource signed exchanges to
track the users by changing content-type and detecting the image loading
failure.

> ## Multiple subresource signed exchanges

Reviewer: This section is covered in the other explainer.

Author: Merged

If there are multiple matching subresource signed exchanges (example:
script.js.sxg and image.jpg.sxg), the UA must check that there is no error in
the all signed exchanges (eg: sig matching, URL matching, Merkle Integrity
error) before processing the content of the signed exchanges. This is intended

> to prevent the distributor from encoding a tracking ID into the set of

Reviewer: And to prevent the distributor from selecting a version of the subresource that isn't compatible with the selected version of the main resource, which might introduce a security hole.

[imagesrcset attribute of preload link](https://github.com/w3c/preload/issues/120).

For Example:
```
Link: <https://publisher.example/wide.jpg>;
      rel="preload";
      as="image";
      imagesrcset="https://publisher.example/wide.jpg 640w,
                   https://publisher.example/narrow.jpg 320w";
      imagesizes="(max-width: 640px) 100vw, 640px"
```
In this case, we want to declare that both wide.jpg and narrow.jpg can be loaded
from signed exchanges and the header-integrity is sha256-XXX and sha256-YYY. But
how to express in this preload link header?

In our proposal, we can have two allowed-alt-sxg link headers.
```
Link: <https://publisher.example/wide.jpg>;
      rel="allowed-alt-sxg";
      header-integrity="sha256-XXX"
Link: <https://publisher.example/narrow.jpg>;
      rel="allowed-alt-sxg";
      header-integrity="sha256-YYY"
```

## Lifetime of the entry in SignedExchangeCache
The UA must check both the
[signature expire time](https://wicg.github.io/webpackage/loading.html#exchange-signature-expiration-time)
of the signed exchange and Cache-Control header of the outer response. The UA
may discard the entry if it is expired.

> ## Content negotiation using Variants and Variant-Key

Reviewer: Only include this section in one of the explainers.

Author: Merged.

        rel="alternate";
        type="application/signed-exchange;v=b3";
        anchor="https://cdn.publisher.example/lib.js"
```

To prevent an attacker from loading an incompatible version of the subresource,
the resource _inside_ the signed exchange has to identify the exact version of
the replacement signed exchange using a `Link: ... rel="allowed-alt-sxg"` with
the hash of the signed headers (which themselves include a hash of the content).

To prevent a tracker from conveying a user ID in their choice of which
subresources to prefetch, the inner resource has to preload the same
subresources that the aggregator prefetches.
And the inner response of the main resource signed exchange (article.html.sxg)
has a preload header and an allowed-alt-sxg header:

```
Link: <https://cdn.publisher.example/lib.js>;
        rel="preload";
        as="script"
Link: <https://cdn.publisher.example/lib.js>;
        rel="allowed-alt-sxg";
        header-integrity="sha256-XXXXXX"
```

The UA recursively prefetches lib.js.sxg.

If the user navigates to the expected article, both the main resource of the
article and the script subresource are loaded from the prefetched signed
exchanges.

> # Proposal

Reviewer: ```suggestion
# Algorithm sketch
```

> - While prenavigating an HTML resource in signed exchange format:

Author: Added a link to the github issue.

    present for the same preload if they include `variants` and `variant-key`
    attributes. In that case, the UA uses the algorithm written in
    [HTTP Representation Variants](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html)
    spec to find the matching header.)
    1. If an `allowed-alt-sxg` link exists, check whether the signed exchange
    was served with a matching “alternate” link HTTP header.
    1. If the outer signed exchange did identify an alternate version of the
    subresource, prefetch the subresource signed exchange instead of prefetching
    the original resource URL.
    1. If the UA successfully prefetches the signed exchange (including the
    merkle integrity check of the body), it stores the parsedExchange which is
    the result of
    [parsing a signed exchange](https://wicg.github.io/webpackage/loading.html#parsing-a-signed-exchange)
    (= inner request URL and inner response) to a new cache mechanism which is
    attached to the Document.
- While
[navigating across documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents),

> the UA copies the parsedExchanges in the cache of the source document to the

Author: Done.

    request's stashed exchange to the matching parsedExchange so the resource
    will be loaded from the cached signed exchange.

# Detailed design discussion
## Header integrity of signed exchange
We propose a new `rel="allowed-alt-sxg"` link header with a `header-integrity`
parameter. Publishers can declare that the subresource can be served by a signed
exchange, using this link header in the inner response of the main resource
signed exchange.

```
Link: <https://cdn.publisher.example/lib.js>;
        rel="allowed-alt-sxg";
        header-integrity="sha256-XXXXXX"
```

This `header-integrity` parameter is the SHA256 hash value of the
*signedHeaders* value from the
[application/signed-exchange format](https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange)
for integrity checking. This signedHeaders is *"the canonical serialization of
the CBOR representation of the response headers of the exchange represented by
the application/signed-exchange resource, excluding the Signature header
field"*. So this value doesn’t change even if the publisher signs the content

> again or changes the signing key. This header-integrity value also guarantees

Reviewer: ```suggestion
again or changes the signing key, but it does change if any of the headers or body change. (It catches changes to the body because a valid signed exchange's headers have to include a `Digest` value that covers the body.)
```

the content body hasn't changed, because the signed headers are required to
include a `Digest` header with a hash of the content body.

> The UA needs to check that this value of the prefetched subresource signed

Reviewer: I *think* this paragraph is already explained elsewhere.

exchange is same as the header-integrity attribute of allowed-alt-sxg link
header in the response from the publisher. This is intended to prevent the
distributor from encoding a tracking ID into the subresource signed exchange.

We can’t use the
[SRI’s integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)
for this purpose, because SRI’s integrity only covers the content body and not
any of the headers. So if the UA use the SRI’s integrity value in

> allowed-alt-sxg link header, we can use the subresource signed exchanges to

Reviewer: ```suggestion
`allowed-alt-sxg` link header, a tracker can use signed exchanges that differ in headers like their `Content-Type` to transfer a user ID via a pattern of parsing failures.
```

track the users by changing content-type and detecting the image loading
failure.

## Multiple subresource signed exchanges
If there are multiple subresource preloads that are provided by prefetches on
the previous page in signed exchange format (example: script.js.sxg and

> image.jpg.sxg), the UA must check that there is no error in the any of the

Reviewer: This "must" is covered above. Is there anything in this section that's not already covered? If there is, try to prune out the repetitive bits by referring to the algorithm. If not, just remove this section.

If we can declare the header-integrity value in the existing preload link HTTP
header, we don’t need to introduce a new "allowed-alt-sxg" link HTTP header.
However, the [imagesrcset attribute](https://github.com/w3c/preload/issues/120)
allows a single preload link to declare multiple different target URLs, and it's
difficult to embed a `header-integrity` value for each of those URLs into the
existing syntax. Instead, we use a separate link that gives a hash of the
expected content for each of the possible URLs, while the preload tag continues
to select which of the URLs is actually used.

For Example:
```
Link: <https://publisher.example/wide.jpg>;
      rel="preload";
      as="image";
      imagesrcset="https://publisher.example/wide.jpg 640w,
                   https://publisher.example/narrow.jpg 320w";
      imagesizes="(max-width: 640px) 100vw, 640px"
Link: <https://publisher.example/wide.jpg>;
      rel="allowed-alt-sxg";
      header-integrity="sha256-XXX"
Link: <https://publisher.example/narrow.jpg>;
      rel="allowed-alt-sxg";
      header-integrity="sha256-YYY"
```

## Lifetime of the entry in SignedExchangeCache
The UA must check both the

> [signature expire time](https://wicg.github.io/webpackage/loading.html#exchange-signature-expiration-time)

Reviewer: ```suggestion
[signature expiration time](https://wicg.github.io/webpackage/loading.html#exchange-signature-expiration-time)
```

          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/jpeg";
          anchor="https://publisher.example/image";
    Link: <https://feed.example/publisher.example/image_webp.sxg>;
          rel="alternate";
          type="application/signed-exchange;v=b3";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/webp";
          anchor="https://publisher.example/image";
    ```
- In **inner** response header of article.html.sxg:
    ```
    Link: <https://publisher.example/image>;
          rel="allowed-alt-sxg";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/jpeg";
          header-integrity="sha256-AAA"
    Link: <https://publisher.example/image>;
          rel="allowed-alt-sxg";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/webp";
          header-integrity="sha256-BBB"
    Link: <https://publisher.example/image>; rel=preload; as=image;
    ```
If a UA supports WebP, the UA must prefetch **image_webp.sxg** holding a WebP
image. Otherwise the UA must prefetch **image_jpeg.sxg** holding a JPEG image.

> ## Security and Privacy Considerations

Reviewer: ```suggestion
# Security and Privacy Considerations
```

- The publishers can know whether the referrer page has prefetched the signed
exchange subresources or not by checking the resource timing information. But
this only exposes 1 bit information (= yes or no) because UAs can use the cached
signed exchange only if the required signed exchanges are all available.
- To prevent tracking (user ID transfer), if the aggregator failed to prefetch a
subresource that the main resource preloads, the UA must drop all of the
subresource prefetches. If the aggregator prefetches a superset of the preloaded
subresources, the UA must drop the ones that weren't preloaded.
- The UA can use the prefetched signed exchange subresources, only when they

> were prefetched in the rimmediate eferrer page. This is intended to avoid

Reviewer: ```suggestion
were prefetched in the immediate referrer page. This is intended to avoid
```

leaking the prefetching state to succeeding pages.

> - The UA check the header integrity value, so the distributor of the subresource

Reviewer: ```suggestion
- The UA checks the header integrity value, so the distributor of the subresource
```

signed exchange can’t inject arbitrary resources to the publisher’s page. This

> prevents distributors from sending tracking IDs to the publisher’s page.

Reviewer: ```suggestion
prevents distributors from sending mismatched versions or tracking IDs to the publisher’s page.
```

- The UA needs to check that the signed
[request URL](https://wicg.github.io/webpackage/loading.html#exchange-request-url)
matches the preload link and not just that the header-integrity value matches,
since the header-integrity hash doesn't cover the request URL.

> - The UA must fetch the signed exchange subresource

Reviewer: It's more that the UA's decision of whether to fetch signed exchange subresources MUST NOT depend on whether the HTTP cache already contains the original subresource. The UA is free to avoid fetching them for other reasons.

(https://cdn.feed.example/cdn.publisher.example/lib.js.sxg) while processing the
prefetch link element  (`<link rel="prefetch"
href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`)
even if the HTTP cache already contains the original subresource
(http://cdn.publisher.example/lib.js). Otherwise it leaks the state of
publisher’s site to the distributor of the signed exchange.
- If a replaced subresource prefetch hasn't completed by the time the UA would
start fetching it in the course of loading the next page, the UA must cancel
that prefetch and fetch the resource from its original URL. This prevents the

> distributor from interfering the publisher’s page. (Eg. intentionally block or

Reviewer: ```suggestion
distributor from interfering the publisher’s page. (Eg. intentionally blocking or delaying...
```

    - No
1. What data does this specification expose to an origin? Please also document
what data is identical to data exposed by other features, in the same or
different contexts.
    - The existence of the alternative signed exchange in HTTP Cache is exposed.
    - But this is the same as the existing behavior when directly prefetching the
    signed exchange using `<link rel=prelfetch>`.
1. Does this specification enable new script execution/loading mechanisms?
    - This specification introduces a new script loading path, from prefetched
    signed exchange. The existing security checks such as CSP/CORP must be
    applied as if the script is loaded from the original URL.
1. Does this specification allow an origin to access other devices?
    - No
1. Does this specification allow an origin some measure of control over a user
agent’s native UI?
    - No
1. What temporary identifiers might this this specification create or expose to
the web?
    - None
1. How does this specification distinguish between behavior in first-party and
third-party contexts?
    - This feature treats all entities (the distributor of signed exchange, the

>     publisher site, the origin of subresource URL) as third-party origins.

Reviewer: I think the publisher's site winds up treated as a first-party origin, since it gets to be top-level after the navigation.

The behavior doesn't line neatly up under the heading of first- vs third-party, but I think we're trying to treat the distributor as a third-party, and of course cross-origin subresources are third-parties.

Author: Done.

resource from the publisher's server, this
[privacy leak problem](https://wicg.github.io/webpackage/draft-yasskin-wpack-use-cases.html#private-prefetch)
 can't be solved.

Signed Exchange solve this privacy-preserving prefetching problem for main
resources. If the publisher is providing the article in signed exchange format
(article.html.sxg), the UA can prefetch the signed exchange from the feed site’s
own server while the user is browsing the feed site. But there is no way to
prefetch subresources in a privacy-preserving manner yet.

Our proposal can solve this problem for subresources:
1. The publisher provides the script file in signed exchange format
(lib.js.sxg).
1. The UA prefetches the signed exchange of the script from the feed site’s own
server while the user is browsing the feed site.
1. The HTTP response of the article from the publisher's server
(https://publisher.example/article.html) has an `allowed-alt-sxg` link header to
declare exactly what version of the subresource is compatible with this main
resource.
   ```
   Link: <https://cdn.publisher.example/lib.js>;
         rel="allowed-alt-sxg";
         header-integrity="sha256-..."
   ```

> ### Recursive subresource signed exchange prefetch

Author: I have one question.

I think "allowed-alt-sxg" is useful even when the HTML is not served via signed exchange.

1. An aggregator site prefetches "https://aggregator.example/publisher.example/lib.js.sxg".
2. The user clicks a link to "https://publisher.example/article.html".
3. The HTTP response header has `Link: <https://publisher.example/lib.js>; rel="allowed-alt-sxg"; header-integrity="sha256-..."` header.
4. When the UA detects <script src="./lib.js">, the UA checks the header-integrity attribute value of the link rel="allowed-alt-sxg" header and the header-integrity value of "lib.js.sxg" in cache:
   - If they match, the UA will use "lib.js.sxg" in the cache.
   - Otherwise, the UA will fetch "https://publisher.example/lib.js".

Do you think this use case is not useful?
If so, I will remove this use case and make "allowed-alt-sxg" work only when the HTML is served via signed exchange.

> ### Recursive subresource signed exchange prefetch

Reviewer: I feel like that's unlikely to be helpful: the referrer has to guess that a page with a particular set of preloaded and replaceable subresources will be the next link clicked, without being able to predict exactly which link it is? I do agree that it's technically possible.

> ### Recursive subresource signed exchange prefetch

Author: OK. I agree with you.
I will remove this use case, and merge "Privacy-preserving prefetching subresources" and "Recursive subresource signed exchange prefetch".

To prevent a tracker from conveying a user ID in their choice of which
subresources to prefetch, the inner resource has to preload the same
subresources that the aggregator prefetches.
And the inner response of the main resource signed exchange (article.html.sxg)
has a preload header and an allowed-alt-sxg header:

```
Link: <https://cdn.publisher.example/lib.js>;
        rel="preload";
        as="script"
Link: <https://cdn.publisher.example/lib.js>;
        rel="allowed-alt-sxg";
        header-integrity="sha256-XXXXXX"
```

The UA recursively prefetches lib.js.sxg.

If the user navigates to the expected article, both the main resource of the
article and the script subresource are loaded from the prefetched signed
exchanges.

# Proposal

## Privacy-preserving prefetching subresources

1. Introduce two new cache mechanism SignedExchangeCache and
PrefetchedSignedExchangeCache in a Document.
   - SignedExchangeCache keeps the signed exchanges which are
prefetched from the Document.
   - PrefetchedSignedExchangeCache keeps the signed
exchanges which were passed from the referrer Document which triggered the
navigation to the current Document.
1. While processing

> "[prefetch](https://html.spec.whatwg.org/multipage/links.html#link-type-prefetch)"

Author: Done.

link (eg: `<link rel="prefetch"
href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`):

>    - If succeeded prefetching the signed exchange (including merkle integrity

Author: Done.

>    check of the body), stores the parsedExchange which is the result of

Author: Done.

>    [parsing a signed exchange](https://wicg.github.io/webpackage/loading.html#ref-for-parsing-a-signed-exchange)

Author: Done.

   (= inner request URL and inner response) to the SignedExchangeCache.
1. While
[navigating across documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents),

> copy the parsedExchanges in the SignedExchangeCache of the source document to

Reviewer: Presumably except for the one that serves the navigation itself?

Author: Yes.
Added.

the PrefetchedSignedExchangeCache of the target document. This is intended to
provide a way to pass the cached signed exchange across origins even if the UA

> is using origin isolated HTTPCache mechanism.

Author: Added
> (Note that [header integrity check](#header-integrity-of-signed-exchange) prohibits the distributor from passing a tracking ID to the publisher.)

> 1. While processing

Reviewer: This step should set up what it's trying to do. I think that's something like

> The navigated-to document has a set of preloads for which it uses the `allowed-alt-sxg` link relation to declare that they can be served by signed exchanges. The UA either serves all of them from SXGs prefetched by the previous page, or none of them.

Author: Added.

[preload](https://html.spec.whatwg.org/multipage/links.html#link-type-preload)
link HTTP headers (eg: Link: <https://cdn.publisher.example/lib.js>;
rel="preload"; as="script"):
   - Check whether matching "allowed-alt-sxg" link HTTP header
   (`Link: <https://cdn.publisher.example/lib.js>;rel="allowed-alt-sxg";header-integrity="sha256-.."`)
   exists or not. (Note that if the allowed-alt-sxg link HTTP header has
   variants and variant-key attributes, the UA must execute the algorithm
   written in
   [HTTP Representation Variants spec](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html)
   to find the matching header.)
   - Check whether all preload links which have matching allowed-alt-sxg link
   header have matching (url and header-integrity) parsedExchange in

>    PrefetchedSignedExchangeCache. If the check passes, set request's stashed

Reviewer: "request" hasn't been introduced yet.

Author: Changed to refer "linked resource fetch setup steps".

   exchange to the parsedExchange so the resource will be loaded from the cached
   signed exchange.
   - After processing the link HTTP headers, clears the
   PrefetchedSignedExchangeCache.

> ## Recursive subresource signed exchange prefetch

Reviewer: This also looks like the same scenario as the previous section.

Author: Removed "Privacy-preserving prefetching subresources".

While prefetching an HTML resource in signed exchange format:

1. When the UA detects a "preload" link HTTP header in the inner response, check
whether a matching “allowed-alt-sxg” link HTTP header in the inner response
exists or not. (Note that multiple `allowed-alt-sxg` links can be present for
the same preload if they include `variants` and `variant-key` attributes. In
that case, the UA uses the algorithm written in
[HTTP Representation Variants](https://httpwg.org/http-extensions/draft-ietf-httpbis-variants.html)
spec to find the matching header.)
1. If an `allowed-alt-sxg` link exists, check whether the signed exchange was
served with a matching “alternate” link HTTP header.
1. If the outer signed exchange did identify an alternate version of the
subresource, prefetch the subresource signed exchange instead of prefetching the
original resource URL.
1. The prefetched signed exchange will be stored to the SignedExchangeCache of
the Document. And it will be passed to the next Document and used while
processing the preload link header. This behavior is written in above.

# Detailed design discussion
## Header integrity of signed exchange

> The `header-integrity` parameter of the `allowed-alt-sxg` link is the SHA256

Reviewer: There should probably be a section that starts with a definition of the `allowed-alt-sxg` link relation, including an example header expressing it.

Author: Done.
How about this?

the application/signed-exchange resource, excluding the Signature header
field"*. So this value doesn’t change even if the publisher signs the content
again or changes the signing key. This header-integrity value also guarantees
the content body hasn't changed, because the signed headers are required to
include a `Digest` header with a hash of the content body.

The UA needs to check that this value of the prefetched subresource signed
exchange is same as the header-integrity attribute of allowed-alt-sxg link
header in the response from the publisher. This is intended to prevent the
distributor from encoding a tracking ID into the subresource signed exchange.

We can’t use the
[SRI’s integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)
for this purpose, because SRI’s integrity only covers the content body and not
any of the headers. So if the UA use the SRI’s integrity value in
allowed-alt-sxg link header, we can use the subresource signed exchanges to
track the users by changing content-type and detecting the image loading
failure.

## Multiple subresource signed exchanges

> If there are multiple matching subresource signed exchanges (example:

Reviewer: It's not clear what "matching" means here. I think you're referring to subresource preloads that are provided by prefetches on the previous page?

Author: Changed to
> If there are multiple subresource preloads that are provided by prefetches on
the previous page in signed exchange format

script.js.sxg and image.jpg.sxg), the UA must check that there is no error in

> the all signed exchanges (eg: sig matching, URL matching, Merkle Integrity

Reviewer: Maybe

```suggestion
any of the signed exchanges (eg: sig matching, URL matching, Merkle Integrity
```

?

Author: Done

error) before processing the content of the signed exchanges. This means that

> the UA can use the subresource signed exchanges only when they are defined in

Reviewer: I don't follow "defined in the header".

selected version of the main resource, which might introduce a security hole.

## Can’t we have a global SignedExchangeCache?
If we can have a global SignedExchangeCache, we can use the all signed exchanges
which have previously prefetched, even if they are not prefetched by the
referrer page. This can improve the performance. But this introduces privacy
issues such as
[Timing Leaks](https://hillbrad.github.io/sri-addressable-caching/sri-addressable-caching.html#cache_timing)
and [Deterministic History Leaks](https://hillbrad.github.io/sri-addressable-caching/sri-addressable-caching.html#history_leaks)
described in [the document about Subresource Integrity Addressable Caching](https://hillbrad.github.io/sri-addressable-caching/sri-addressable-caching.html).

To avoid these privacy issues, we introduced the limitation that we can only use
the signed exchanges which were prefetched by the referrer page.

## Can’t we merge allowed-alt-sxg to preload header?
If we can declare the header-integrity value in the existing preload link HTTP
header, we don’t need to introduce a new "allowed-alt-sxg" link HTTP header. But

> it becomes complicated when supporting the

Reviewer: This might be easier to read as something like, "However, the imagesrcset attribute allows a single preload link to declare multiple different target URLs, and it's difficult to embed a `header-integrity` value for each of those URLs into the existing syntax. Instead, we use a separate link that gives a hash of the expected content for each of the possible URLs, while the preload tag continues to select which of the URLs is actually used."

(I do want to double-check that you can defend that claim that it's difficult to embed. Why couldn't it just come after the XXXw field?)

    ```
    Link: <https://feed.example/publisher.example/image_jpeg.sxg>;
          rel="alternate";
          type="application/signed-exchange;v=b3";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/jpeg";
          anchor="https://publisher.example/image";
    Link: <https://feed.example/publisher.example/image_webp.sxg>;
          rel="alternate";
          type="application/signed-exchange;v=b3";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/webp";
          anchor="https://publisher.example/image";
    ```
- In **inner** response header of article.html.sxg:
    ```
    Link: <https://publisher.example/image>;
          rel="allowed-alt-sxg";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/jpeg";
          header-integrity="sha256-AAA"
    Link: <https://publisher.example/image>;
          rel="allowed-alt-sxg";
          variants-05="accept;image/jpeg;image/webp";
          variant-key-05="image/webp";
          header-integrity="sha256-BBB"
    Link: <https://publisher.example/image>; rel=preload; as=image;
    ```

> If a UA supports WebP, the UA should prefetch **image_webp.sxg** which content

Reviewer: ```suggestion
If a UA supports WebP, the UA must prefetch **image_webp.sxg** holding a WebP image.
```

This isn't "should", since it's defined by the variants algorithm.

is WebP format. Otherwise the UA should prefetch **image_jpeg.sxg** which
content is JPEG format.

## Security and Privacy Considerations
- The publishers can know whether the referrer page has prefetched the signed
exchange subresources or not by checking the resource timing information. But
this only exposes 1 bit information (= yes or no) because UAs can use the cached
signed exchange only if the required signed exchanges are all available.
- The UA can use the prefetched signed exchange subresources, only when they

> were prefetched in the referrer page. This is intended to avoid leaking the

Reviewer: ```suggestion
were prefetched in the immediate referrer page. This is intended to avoid leaking the
```

It's not clear to me exactly what information gets leaked if we do allow a series of navigations, especially same-origin navigations, to continue using the same set of prefetches. Maye the benefit isn't there, since the first page can just re-prefetch the same-origin content, but then we should say that instead of being vague about the leak.

prefetching state to succeeding pages.
- The UA check the header integrity value, so the distributor of the subresource
signed exchange can’t inject arbitrary resources to the publisher’s page. This
prevents distributors from sending tracking IDs to the publisher’s page.
- The UA need to check both
[the request URL](https://wicg.github.io/webpackage/loading.html#exchange-request-url)

> and the header integrity value of the signed exchange to avoid

Reviewer: "The UA needs to check that the signed request URL matches the preload link and not just that the header-integrity value matches, since the header-integrity hash doesn't cover the request URL."

[Origin Laundering](https://hillbrad.github.io/sri-addressable-caching/sri-addressable-caching.html#origin_laundering)
attack. The URL is not in signedHeaders, so the header integrity value can’t
check the URL is correct.
- The UA must fetch the signed exchange subresource
(https://cdn.feed.example/cdn.publisher.example/lib.js.sxg) while processing the
prefetch link element  (`<link rel="prefetch"
href="https://cdn.feed.example/cdn.publisher.example/lib.js.sxg" as="script">`)

> even if there is the original subresource (http://cdn.publisher.example/lib.js)

Reviewer: ```suggestion
even if the HTTP cache already contains the original subresource (http://cdn.publisher.example/lib.js)
```

Although it's not clear to me in this system how the UA would even know what the original subresource URL is before fetching the SXG.

Author: Done.

If we don't need to think about privacy, when the HTTP cache has "lib.js", UAs don't need to fetch "lib.js.sxg" when the UA see rel=preload and rel=allowed-alt-sxg and rel=alternate for better performance. But we can't do that for the privacy reason.

in the HTTPCache. Otherwise it leaks the state of  publisher’s site to the
distributor of the signed exchange.

> - The UA must fetch the alternate signed exchange subresource (lib.js.sxg) even

Reviewer: This is a duplicate of the previous bullet point.

if there is the original subresource (lib.js) in the HTTPCache. Otherwise it
leaks the state of publisher’s site to the distributor of the signed exchange.

> - To prevent tracking (user ID transfer), if the aggregator failed to prefetch a

Reviewer: This is very similar to the first bullet in this list. Put them next to each other.

subresource that the main resource preloads, the UA must drop all of the
subresource prefetches. If the aggregator prefetches a superset of the preloaded
subresources, the UA must drop the ones that weren't preloaded.

> - The UA must ignore the not-yet-completed subresource signed exchange prefetch

Reviewer: This one should probably start, "If a replaced subresource prefetch hasn't completed by the time the UA would start fetching it in the course of loading the next page, the UA must cancel that prefetch and fetch the resource from its original URL. This prevents the distributor from ..."

to prevent distributors from interfering the publisher’s page. (Eg.
Intentionally block or delay the subresource loading.) So if the user clicks the
link to the article while the UA is still prefetching the subresource signed
exchange, the UA must fetch the subresource from the original URL after
navigation.
- The UA can use the prefetched signed exchange subresources only if the preload
HTTP link headers for the subresources exist in the response headers. It is

> because the UA need to check the availability of all required signed exchanges

Reviewer: Isn't this the same as the "To prevent tracking (user ID transfer)" item?

before start loading subresources.

## [Self-Review Questionnaire: Security and Privacy](https://www.w3.org/TR/security-privacy-questionnaire/)
1. What information might this feature expose to Web sites or other parties, and
for what purposes is that exposure necessary?

>    - The existence of the alternative signed exchange in HTTP Cache is exposed.

Reviewer: Is this true? Who is it exposed to and under what circumstances? When is the alternative SXG even in the HTTP cache? Which partition?

1. Is this specification exposing the minimum amount of information necessary to
power the feature?

>    - Yes. This proposal has limitations such as "all subresource SXG must be

Reviewer: Say what the limitations accomplish here, not just that some phrases exist in other parts of this document.

   finished prefetching, otherwise ignored", "subresource SXG must be prefetched
   even if the original subresource exists in HTTPCache".
1. How does this specification deal with personal information or
personally-identifiable information or information derived thereof?

>    - Signed Exchange should not include personal information.

Reviewer: Also, any personal information that was incorrectly included in a signed exchange would be the information of the aggregator that fetched the SXG, and not the end user.

The use of `<link rel=alternate>` to identify the SXG for the current page could inform the UA to omit credentials in fetching that SXG, which would prevent any personal information from being accidentally included.
